File: so/lab2/mycp.c
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#define BUFFER_SIZE 8
char buffer[BUFFER_SIZE];

#define MINIMUM(a, b) ((a) < (b) ? (a) : (b))

// Computes the length of a nul-terminated string
size_t strlen(const char* s) {
    size_t len = 0;
    while (*s++) {
        ++len;
    }
    return len;
}

// Prints an error message
void print_error(const char* message) {
    write(STDERR_FILENO, message, strlen(message));
    write(STDERR_FILENO, "\n", 1);
}

int main(int argc, char** argv) {
    // Check arguments
    if (argc != 3) {
        print_error("Wrong parameter number");
        return 1;
    }

    const char* in_path = argv[1];
    const char* out_path = argv[2];

    // Open the original input file
    int in_fd = open(in_path, O_RDONLY);
    if (in_fd < 0) {
        print_error("Cannot open input file");
        return 1;
    }

    // Create a new output file
    int out_fd = creat(out_path, S_IRUSR | S_IWUSR);
    if (out_fd < 0) {
        print_error("Could not create output file");
        return 1;
    }

    // Determine how big the file is
    struct stat info;
    if (fstat(in_fd, &info) < 0) {
        print_error("Could not get file size");
        return 1;
    }

    const off_t file_size = info.st_size;

    // Data copy loop
    off_t xferred = 0;
    ssize_t read_result = 0;
    int has_error = 0;

    while (xferred < file_size) {
        const size_t left_to_read = file_size - xferred;
        const size_t to_read_now = MINIMUM(BUFFER_SIZE, left_to_read);

        // Try to read into the buffer
        read_result = read(in_fd, buffer, to_read_now);
        if (read_result < 0) {
            print_error("Error reading file");
            has_error = 1;
            break;
        }

        size_t written = 0;
        ssize_t write_result = 0;
        while (written < read_result) {
            const size_t left_to_write = read_result - written;
            write_result = write(out_fd, buffer, left_to_write);

            if (write_result < 0) {
                break;
            }

            written += write_result;
        }

        if (write_result < 0) {
            print_error("Error writing buffer to disk");
            has_error = 1;
            break;
        }

        xferred += read_result;
    }

    // Close file descriptors
    if (close(in_fd) < 0) {
        print_error("Could not close input file descriptor");
        has_error = 1;
    }

    if (close(out_fd) < 0) {
        print_error("Could not close output file descriptor");
        has_error = 1;
    }

    if (has_error) {
        print_error("File not fully written");
        return 1;
    }
}

File: so/lab2/src/bin/mycp.rs
use std::{cmp, env, mem, result};

type Result<T> = result::Result<T, libc::c_int>;
type FileDescriptor = libc::c_int;

enum FileMode {
    ReadOnly,
    Create,
}

struct File {
    fd: FileDescriptor,
}

impl File {
    /// Opens a file and returns a descriptor to it
    fn open(path: &str, mode: FileMode) -> Result<Self> {
        let path = path.as_ptr() as *const libc::c_char;

        let result = unsafe {
            match mode {
                FileMode::ReadOnly => libc::open(path, libc::O_RDONLY),
                FileMode::Create => libc::creat(path, libc::S_IRUSR | libc::S_IWUSR),
            }
        };

        match result {
            error if result < 0 => Err(error),
            fd => Ok(Self { fd }),
        }
    }

    /// Returns the size, in bytes, of a given file.
    fn size(&self) -> Result<u64> {
        let mut info = mem::MaybeUninit::uninit();
        let result = unsafe { libc::fstat(self.fd, info.as_mut_ptr()) };

        match result {
            error if result < 0 => Err(error),
            _ok => unsafe { Ok(info.assume_init().st_size as u64) },
        }
    }

    /// Reads from a given file into a in-memory buffer
    fn read(&self, buffer: &mut [u8], how_much_to_read: usize) -> Result<u64> {
        let buffer = buffer.as_mut_ptr() as *mut _;
        let result = unsafe { libc::read(self.fd, buffer, how_much_to_read) };

        match result {
            error if result < 0 => Err(error as i32),
            ok => Ok(ok as u64),
        }
    }

    /// Writes from an input buffer to the file
    fn write(&self, buffer: &[u8], how_much_to_write: usize) -> Result<u64> {
        let buffer = buffer.as_ptr() as *const _;
        let result = unsafe { libc::write(self.fd, buffer, how_much_to_write) };

        match result {
            error if result < 0 => Err(error as i32),
            ok => Ok(ok as u64),
        }
    }
}

impl Drop for File {
    fn drop(&mut self) {
        if unsafe { libc::close(self.fd) } != 0 {
            panic!("Failed to close file");
        }
    }
}

fn main() {
    let (src, dst) = {
        let mut args = env::args();
        let program_name = args.next().unwrap();

        if args.len() != 2 {
            panic!(
                "Wrong number of parameters passed!\nUsage:\n\t{} input_path output_path",
                program_name
            );
        }

        // Open the original input file
        let src_path = args.next().unwrap();
        let src = File::open(&src_path, FileMode::ReadOnly).expect("Failed to open input file");

        // Create a new output file
        let dst_path = args.next().unwrap();
        let dst = File::open(&dst_path, FileMode::Create).expect("Failed to create output file");

        (src, dst)
    };

    const BUFFER_SIZE: usize = 8;
    let mut buffer = [0u8; BUFFER_SIZE];

    let file_size = src.size().expect("Could not determine input file size");

    let mut xferred = 0u64;

    // Data copy loop
    while xferred < file_size {
        let left_to_copy = file_size - xferred;
        let max_read_size = cmp::min(BUFFER_SIZE, left_to_copy as usize);

        // Fill the buffer
        let read_size = src
            .read(&mut buffer, max_read_size)
            .expect("Failed to read from input file");

        let mut written = 0;
        while written < read_size {
            let left_to_write = read_size - written;
            let write_size = dst
                .write(&buffer, left_to_write as usize)
                .expect("Failed to write to destination file");

            written += write_size;
        }

        xferred += read_size;
    }
}
